src/replication.c

```diff
@@ -967,6 +900,77 @@ update_replica_entry(spec_t *spec, replica_t *replica, int iofd, char *replicaip
 	return replica;
 }
 
+static void
+cleanup_replica(replica_t *replica)
+{
+	MTX_LOCK(&replica->r_mtx);
+	if(replica->mgmt_fd != -1)  {
+		close(replica->mgmt_fd);
+		replica->mgmt_fd = -1;
+	}
+
+	if(replica->iofd != -1)  {
+		close(replica->iofd);
+		replica->iofd = -1;
+	}
+
+	if(replica->sender_epfd != -1) {
+		close(replica->sender_epfd);
+		replica->sender_epfd = -1;
+	}
+
+	if(replica->io_resp_hdr) {
+		free(replica->io_resp_hdr);
+	}
+
+	if(replica->ip) {
+		free(replica->ip);
+	}
+
+	free(replica->mgmt_io_resp_hdr);
+	if (replica->mgmt_io_resp_data) {
+		free(replica->mgmt_io_resp_data);
+		replica->mgmt_io_resp_data = NULL;
+	}
+
+	MTX_UNLOCK(&replica->r_mtx);
+
+	pthread_mutex_destroy(&replica->r_mtx);
+	pthread_cond_destroy(&replica->r_cond);
+	free(replica);
+}
+
+static int
+send_replica_handshake_query(replica_t *replica, spec_t *spec)
+{
+	zvol_io_hdr_t *rmgmtio = NULL;
+	size_t data_len = 0;
+	uint8_t *data;
+	zvol_op_code_t mgmt_opcode = ZVOL_OPCODE_HANDSHAKE;
+	mgmt_cmd_t *mgmt_cmd;
+
+	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
+
+	data_len = strlen(spec->lu->volname) + 1;
+
+	build_replica_mgmt_hdr();
+
+	data = (uint8_t *)malloc(data_len);
+	snprintf((char *)data, BUFSIZE, "%s", spec->lu->volname);
+
+	mgmt_cmd->io_hdr = rmgmtio;
+	mgmt_cmd->io_bytes = 0;
+	mgmt_cmd->data = data;
+	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+	MTX_LOCK(&replica->r_mtx);
+	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+	MTX_UNLOCK(&replica->r_mtx);
+
+	handle_write_data_event(replica);
```

Comment:
we don't need this. we get a EPOLLOUT in the starting.

Reply:
ok

Reply:
sorry.. this comment is at wrong place.. we need to call this fn here..
I wanted to comment at the place where handshake command was made